src/main/resources/testchipip/csrc/mm.cc

```diff
@@ -0,0 +1,150 @@
+// See LICENSE for license details.
+
+#include "mm.h"
+#include <iostream>
+#include <fstream>
+#include <cstdlib>
+#include <cstring>
+#include <cassert>
+#include <sys/mman.h>
+
+void mm_t::write(uint64_t addr, uint8_t *data, uint64_t strb, uint64_t size)
+{
+  strb &= ((1 << size) - 1) << (addr % word_size);
+  addr %= this->size;
+
+  uint8_t *base = this->data + (addr / word_size) * word_size;
+  for (int i = 0; i < word_size; i++) {
+    if (strb & 1)
+      base[i] = data[i];
+    strb >>= 1;
+  }
+}
+
+std::vector<char> mm_t::read(uint64_t addr)
+{
+  addr %= this->size;
+
+  uint8_t *base = this->data + addr;
+  return std::vector<char>(base, base + word_size);
+}
+
+void mm_t::init(size_t sz, int wsz, int lsz)
+{
+  assert(wsz > 0 && lsz > 0 && (lsz & (lsz-1)) == 0 && lsz % wsz == 0);
+  word_size = wsz;
+  line_size = lsz;
+  data = (uint8_t *) mmap(
+          NULL, sz, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_ANONYMOUS, -1, 0);
+  size = sz;
+}
+
+mm_t::~mm_t()
+{
+  munmap(data, this->size);
+}
+
+void mm_magic_t::init(size_t sz, int wsz, int lsz)
```

Comment:
Why do we need to reimplement a magic AXI memory?

Reply:
Per chipyard meeting discussion. Simulating magic memory in a C++ model is faster than simulating SimAXIMem RTL.